### PR TITLE
[WIP] Check whether a tweet is scraped

### PR DESCRIPTION
**Pull request by AnshulMalik**
Issue: #1353 

### Short description
Initial steps to add the new feature
- Created a new servlet to handle new endpoint `check.json`

### CODE REVIEW

**Comment by vibhcool** [comment]
@AnshulMalik, what is the possible usecase for checking if tweet is scraped? It can be checked on quering from cache. there is a existing method to check if tweet exist. 
the issue is to scrape tweets and output as archive from search endpoint.

**Comment by vibhcool** [comment]
you may discuss in issue or gitter, how to approach for the solution before implementing it :)

File: src/org/loklak/LoklakServer.java [hunk 1]
```diff
@@ -104,25 +104,7 @@
...
-import org.loklak.api.search.ShortlinkFromTweetServlet;
-import org.loklak.api.search.SuggestServlet;
-import org.loklak.api.search.TimeAndDateService;
-import org.loklak.api.search.TweetScraper;
-import org.loklak.api.search.UserServlet;
-import org.loklak.api.search.VideoUrlService;
-import org.loklak.api.search.WeiboUserInfo;
-import org.loklak.api.search.WikiGeoData;
-import org.loklak.api.search.WordpressCrawlerService;
+import org.loklak.api.search.*;
```
Comment (sudheesh001) [inline src/org/loklak/LoklakServer.java]: No! We didn't do this for a reason. Please undo this change.

File: src/org/loklak/api/search/TweetCheckServlet.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        String tweetQuery = query.get("tweet", "");
+        String userQuery = query.get("user", "");
+
+        response.setCharacterEncoding("UTF-8");
+
+        PrintWriter sos = response.getWriter();
+
+        JSONObject result = new JSONObject(true);
+
+        if(tweetQuery != null) {
```
Comment (sudheesh001) [inline src/org/loklak/api/search/TweetCheckServlet.java]: space consistency. `if (` instead of `if(`

File: src/org/loklak/api/search/TweetCheckServlet.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+            String tweetId = parseTweetId(tweetQuery);
+            boolean exists = DAO.existMessage(tweetId);
+            if(exists) {
```
Comment (sudheesh001) [inline src/org/loklak/api/search/TweetCheckServlet.java]: Same space consistency here.

File: src/org/loklak/api/search/TweetCheckServlet.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            }
+        }
+
+        query.setResponse(response, jsonp ? "application/javascript" : "application/json");
+
+        if (jsonp) sos.print(callback + "(");
+        sos.print(result.toString(minified ? 0 : 2));
+        if (jsonp) sos.println(");");
+        sos.println();
+        sos.flush(); sos.close();
```
Comment (sudheesh001) [inline src/org/loklak/api/search/TweetCheckServlet.java]: `sos.close()` on new line.

File: src/org/loklak/api/search/TweetCheckServlet.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        Query query = RemoteAccess.evaluate(request);
+
+        // manage DoS
+        if (query.isDoS_blackout()) {response.sendError(503, "your request frequency is too high"); return;}
+
+        String callback = query.get("callback", "");
+        boolean jsonp = callback != null && callback.length() > 0;
+        boolean minified = query.get("minified", false);
+
+        String tweetQuery = query.get("tweet", "");
```
Comment (sudheesh001) [inline src/org/loklak/api/search/TweetCheckServlet.java]: Does the user put the entire link here? or just the `tweetid`? If it's the entire link, why is it designed this way? In what scenarios will someone be using this API call with a `full` URL string.

In ideal cases if you were looking up something, you'd look things up by ID and not by the entire path to the ID right, the same should be the analogy here. Making this just the `tweetID:string` will make `parseTweetId()` a redundant method.

**Comment by sudheesh001** [comment]
A little more thought experiment
- [ ] What should the API do when it doesn't find the required string? (Add link to the queue to crawl/scrape; Validate that a tweet exists on twitter but we don't have it indexed yet?)
- [ ] What is the rationale behind using the entire URL in the parameters to be passed? As pointed out in the reviews above.

File: src/org/loklak/LoklakServer.java (hunk 1 above)
Comment (AnshulMalik) [reply src/org/loklak/LoklakServer.java]: Ok but May I know the reason?

File: src/org/loklak/api/search/TweetCheckServlet.java (hunk 5 above)
Comment (AnshulMalik) [reply src/org/loklak/api/search/TweetCheckServlet.java]: The issue said: 
- `/api/show.json?tweet=https://twitter.com/ArtisticBondage/status/885041783079661569`
- `/api/show.json?tweet=885041783079661569`

Now if it's a link, we extract the id, and if it's already id, we are good to go.

So I added this method

File: src/org/loklak/LoklakServer.java (hunk 1 above)
Comment (sudheesh001) [reply src/org/loklak/LoklakServer.java]: The main reason why we do not do this is future compatibility, if there's a package which says `a.b.c.*` and `x.y.z.*` both having an class of say `Person` then there are higher chances for errors and the compilers getting confused. Also using the `*` means you're increasing the lookup time for the compiler while compilation by a small time but the resulting byte code will remain the same anyway.

I agree that having a long list of the imports can be quite intimidating but good IDEs always collapse them and you do not have to worry about what you import either because if there's a class you're using which isn't found in the file as an import, the IDE allows you to press a shortcut and it'll auto import that for you.

Comment (AnshulMalik) [reply src/org/loklak/LoklakServer.java]: Thanks
Tip: This change was done be intellij itself :)

File: src/org/loklak/api/search/TweetCheckServlet.java (hunk 5 above)
Comment (sudheesh001) [reply src/org/loklak/api/search/TweetCheckServlet.java]: The question here would be how would you know from the current method that the `tweet` value is a valid ID or a link. Doing a check for `http` contained in the string would not help here at all. I can pass my string to be `tweet=https://126312356123612/ajsdba/1 which will pass and it'll treat the ID as `1`  but this has major problems because the URL isn't by itself a valid URL.

You need to parse `https://twitter.com/{:user}/status/{:id}` for validation.

**Comment by sudheesh001** [comment]
left a comment at https://github.com/loklak/loklak_server/pull/1527#discussion_r138239339 The PR still doesnot adequately address the issue.
This questions isn't answered too
- [ ] What should the API do when it doesn't find the required string? (Add link to the queue to crawl/scrape; Validate that a tweet exists on twitter but we don't have it indexed yet?)

**Comment by mariobehling** [comment]
@AnshulMalik Any update on this? Would be great to see this merged.

@sudheesh001 If there is no other activity should this be closed?
